typhos/notes.py

```diff
@@ -0,0 +1,189 @@
+    source = NotesSource.USER
+
+    # try env var
+    env_notes_path = os.environ.get(NOTES_VAR)
+    if env_notes_path and Path(env_notes_path).is_file():
+        note_data = get_data_from_yaml(device_name, env_notes_path)
+        if note_data:
+            data = note_data
+            source = NotesSource.ENV
+
+    # try user directory
+    user_data_path = platformdirs.user_data_path() / 'device_notes.yaml'
+    if user_data_path.exists():
+        note_data = get_data_from_yaml(device_name, user_data_path)
+        if note_data:
+            data = note_data
+            source = NotesSource.USER
+
+    return source, data
+
+
+def insert_into_yaml(path: Path, device_name: str, data: dict[str, str]) -> None:
+    try:
+        with open(path, 'r') as f:
+            device_notes = yaml.full_load(f)
+    except Exception as ex:
+        logger.warning(f'unable to open existing device info: {ex}')
+        return
+
+    device_notes[device_name] = data
+
+    try:
+        with open(path, 'w') as f:
+            yaml.dump(device_notes, f)
+    except Exception as ex:
+        logger.warning(f'unable to write device info: {ex}')
+        return
+
+
+def write_notes_data(
+    source: NotesSource,
+    device_name: str,
+    data: dict[str, str]
+) -> None:
+    """
+    Write the notes ``data`` to the specified ``source`` under the key ``device_name``
+
+    Parameters
+    ----------
+    source : NotesSource
+        The source to write the data to
+    device_name : str
+        The device name.  Can also be a component. e.g. device_component_name
+    data : dict[str, str]
+        The notes data.  Expected to contain the 'note' and 'timestamp' keys
+    """
+    if source == NotesSource.USER:
+        user_data_path = platformdirs.user_data_path() / 'device_notes.yaml'
+        insert_into_yaml(user_data_path, device_name, data)
+    elif source == NotesSource.ENV:
+        env_data_path = Path(os.environ.get(NOTES_VAR))
+        insert_into_yaml(env_data_path, device_name, data)
+
+
+class TyphosNotesEdit(QtWidgets.QLineEdit):
+    """
+    A QLineEdit for storing notes for a device.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.editingFinished.connect(self.save_note)
+        self.setPlaceholderText('...')
+        self.edit_filter = utils.FrameOnEditFilter(parent=self)
+        self.setFrame(False)
+        self.setStyleSheet("QLineEdit { background: transparent }")
+        self.setReadOnly(True)
+        self.installEventFilter(self.edit_filter)
+
+        # to be initialized later
+        self.device_name: str = None
+        self.notes_source: Optional[NotesSource] = None
+        self.data = {'note': '', 'timestamp': ''}
+
+    def update_tooltip(self) -> None:
+        if self.notes_source is not None:
+            self.setToolTip(f"({self.data['timestamp']}, {self.notes_source.name}):\n"
+                            f"{self.data['note']}")
```

Comment:
I think we can do slightly better for a default tooltip when no description has been written:
![image](https://github.com/pcdshub/typhos/assets/10647860/0337202a-04f0-4a63-9fd3-fb4684370913)

Maybe just some simple plaintext "Click here to edit the description" if the note is empty?